### PR TITLE
Move DeepCat functions to DeepCat module

### PR DESCRIPTION
**Pull request by WMDE-Fisch**
Includes #67 
Introduce a main method.
Dependencies should now be declared when loading the gadget, see manual. 

Bug: T142838

I am not sure if this completely solves the problem of hoisting PerfektesChaos talks about in https://de.wikipedia.org/wiki/Hilfe_Diskussion:Suche/Deepcat#Richtiges_Gadget . Or if this really would mean declaring methods always before using them. ( since the approach here seems to work fine. )

### CODE REVIEW

**Comment by thiemowmde** [comment]
As far as I can see this is not what he meant. What this patch currently does is making all private methods public, which should not be done. At least not for all of them.

What he means is simply that functions are called before they are declared. ~~JSCS~~ [JSHint does have a rule that reports such cases](http://jshint.com/docs/options/#latedef). The solution is to simply move the `function …( … ) { … }` declarations up and order them as required.

File: DeepCat.js [hunk 1]
```diff
@@ -146,53 +86,53 @@
...
 
 		for( i = 0; i < searchTerms.length; i++ ) {
-			if( matchesDeepCatKeyword( searchTerms[ i ] ) ) {
-				requests.push( getAjaxRequest( searchTerms[ i ], i ) );
+			if( this.matchesDeepCatKeyword( searchTerms[ i ] ) ) {
+				requests.push( this.getAjaxRequest( searchTerms[ i ], i ) );
 			}
 		}
 
-		$.when.apply( this, requests ).done( receiveAjaxResponses );
```
Comment (thiemowmde) [inline DeepCat.js:161]: For example, `receiveAjaxResponses` is used up here …

File: DeepCat.js [hunk 2]
```diff
@@ -146,53 +86,53 @@
...
-	}
+	};
 
 	/**
 	 * Process all the AJAX responses from catgraph-jsonp, modify the search string and
 	 * re-submit the search form.
 	 *
 	 * This function can receive an arbitrary number of parameters.
 	 */
-	function receiveAjaxResponses() {
```
Comment (thiemowmde) [inline DeepCat.js:195]: … but declared down here.

**Comment by WMDE-Fisch** [comment]
Thanks Thiemo for looking into that. 
- I am not sure if it is a bad thing if all the methods are public. Personally I think the structure is clearer that way. But lets wait for other opinions.
- Related to hoisting: Ok I feared that that's the answer. But then I ask myself since when is that a problem? Even in JavaScript everybody kind of codes like that and obviously it works. I think its a better style having the top-down things top-down. But that might be an answer that should be given on the talk page if we don"t address the issue. I would also like to hear opinions on that.

**Comment by thiemowmde** [comment]
Did you followed the link I posted? It helped me understand it: http://jshint.com/docs/options/#latedef. Sure, almost all code works both ways around, but it's considered best practice (in JavaScript) and also makes the live of the JavaScript parser easier.

**Comment by WMDE-Fisch** [comment]
Yeah thanks, I looked at it again. The thing is, we even have jshint checks running on Travis and `latedef` is already in there. Still no complains... so ... I am confused.

**Comment by thiemowmde** [comment]
When I run jshint locally on this repository, with the repositories settings, I get:

```
DeepCat.js: line 149, col 14, 'sendAjaxRequests' was used before it was defined.
DeepCat.js: line 164, col 14, 'getAjaxRequest' was used before it was defined.
DeepCat.js: line 181, col 14, 'getAjaxRequestUrl' was used before it was defined.
DeepCat.js: line 195, col 14, 'receiveAjaxResponses' was used before it was defined.
DeepCat.js: line 231, col 14, 'logAndFinishRequest' was used before it was defined.
DeepCat.js: line 242, col 14, 'getLogRequestUrl' was used before it was defined.
DeepCat.js: line 247, col 14, 'finishDeepCatRequest' was used before it was defined.
DeepCat.js: line 335, col 14, 'createErrorMessage' was used before it was defined.
DeepCat.js: line 342, col 14, 'ajaxSuccess' was used before it was defined.
DeepCat.js: line 347, col 14, 'graphError' was used before it was defined.
DeepCat.js: line 352, col 14, 'ajaxError' was used before it was defined.
DeepCat.js: line 360, col 14, 'fatalAjaxError' was used before it was defined.
DeepCat.js: line 365, col 14, 'substituteSearchRequest' was used before it was defined.
DeepCat.js: line 392, col 14, 'substituteInputValues' was used before it was defined.
DeepCat.js: line 434, col 14, 'matchesDeepCatKeyword' was used before it was defined.
DeepCat.js: line 454, col 14, 'checkErrorMessage' was used before it was defined.
DeepCat.js: line 474, col 14, 'refreshSearchTermMock' was used before it was defined.
DeepCat.js: line 488, col 14, 'addAjaxThrobber' was used before it was defined.
DeepCat.js: line 494, col 14, 'removeAjaxThrobber' was used before it was defined.
DeepCat.js: line 500, col 14, 'addSearchFormHint' was used before it was defined.
DeepCat.js: line 512, col 14, 'addSmallFormHint' was used before it was defined.
DeepCat.js: line 521, col 14, 'hasHintCookie' was used before it was defined.
DeepCat.js: line 525, col 14, 'hideHints' was used before it was defined.
DeepCat.js: line 539, col 14, 'hideSmallHint' was used before it was defined.
DeepCat.js: line 545, col 14, 'disableImeAndSuggestions' was used before it was defined.
DeepCat.js: line 550, col 14, 'enableImeAndSuggestions' was used before it was defined.
DeepCat.js: line 555, col 14, 'getMainSearchFormId' was used before it was defined.
DeepCat.js: line 563, col 14, 'advancedSearchFormIsPresent' was used before it was defined.
DeepCat.js: line 591, col 14, 'removeUnicodeNonPrintables' was used before it was defined.
DeepCat.js: line 602, col 14, 'replaceWhiteSpace' was used before it was defined.
DeepCat.js: line 611, col 14, 'makeHintboxCookieToken' was used before it was defined.
DeepCat.js: line 619, col 14, 'stringFormat' was used before it was defined.
DeepCat.js: line 630, col 14, 'loadMessages' was used before it was defined.

tests/DeepCat.tests.js: line 157, col 79, This character may get silently deleted by one or more browsers.

34 errors
```

**Comment by manicki** [comment]
Thanks @thiemowmde !
This is indeed broken very badly since a while. All apparently due to Travis CI using JSHint 2.8.\* which do not report on those issues. This version restriction was added for a reason while ago but this does not seem needed any more.

I will submit a patch that takes this now silly version restriction. As JSHint now gives 34 errors for current master, to make @WMDE-Fisch's work on this PR easier, I'll do it on top of this PR.
It should only be merged when the PR fixing Travis builds is in-sync and CI checks are passing!

**Comment by manicki** [comment]
no, wait, new JSHint does not complain on anything related to what @WMDE-Fisch changes. Which does not seem to help when it comes to this "hoisting" problem

**Comment by thiemowmde** [comment]
This patch turns everything into public methods instead of using private functions. This new structure is nothing JSHint would complain about. The check only applies to functions, not object structures.

I disagree with this change because most of the functions are nothing but implementation details of this gadget and do not have much use for other user scripts and gadgets. By making all this public every change to any of these methods will be a breaking change, but because it is a gadget and gadgets do not have version numbers there is no way to announce such changes, no change log, no dependency management.

The only way to avoid all this is to make the public interface as small as possible.

File: README.md [hunk 3]
```diff
@@ -24,8 +24,10 @@ importStylesheet(  'User:USERNAME/Gadgets/DeepCat.css' );
 
 The current official release of the gadget can be found on wikipedia.org 
 ```
-mw.loader.load( "//de.wikipedia.org/w/index.php?title=User:Christoph Fischer (WMDE)/Gadgets/DeepCat.js&action=raw&ctype=text/javascript" );
-mw.loader.load( "//de.wikipedia.org/w/index.php?title=User:Christoph Fischer (WMDE)/Gadgets/DeepCat.css&action=raw&ctype=text/css" , "text/css" );
+$.when( mw.loader.using( [ 'mediawiki.api.messages', 'mediawiki.jqueryMsg' ] ), $.ready ).done( function() {
+	mw.loader.load( "//de.wikipedia.org/w/index.php?title=User:Christoph Fischer (WMDE)/gadget-test.js&action=raw&ctype=text/javascript" );
```
Comment (manicki) [inline README.md:28]: any reason not to use single quotes in this line and the line below? Just to promote the use of single quotes all over the world!

Comment (WMDE-Fisch) [reply README.md:28]: I might fix that later in following patches ;-)

**Comment by WMDE-Fisch** [comment]
@thiemowmde convinced me, I uploaded a new PR just addressing the hoisting. I will add other patches that build upon it. see #70

**Pull request closed without merging**